src/BizHawk.Emulation.Common/zstd/Zstd.cs

```diff
@@ -0,0 +1,505 @@
+﻿using System;
+using System.IO;
+using System.Runtime.InteropServices;
+
+using BizHawk.BizInvoke;
+using BizHawk.Common;
+
+namespace BizHawk.Emulation.Common
+{
+	public sealed class Zstd : IDisposable
+	{
+		private sealed class ZstdCompressionStreamContext : IDisposable
+		{
+			public readonly IntPtr Zcs;
+
+			public readonly byte[] InputBuffer;
+			public readonly GCHandle InputHandle;
+			public LibZstd.StreamBuffer Input;
+
+			public readonly byte[] OutputBuffer;
+			public readonly GCHandle OutputHandle;
+			public LibZstd.StreamBuffer Output;
+
+			// TODO: tweak these sizes
+
+			// 4 MB input buffer
+			public const int INPUT_BUFFER_SIZE = 1024 * 1024 * 4;
+			// 1 MB output buffer
+			public const int OUTPUT_BUFFER_SIZE = 1024 * 1024 * 1;
+
+			public bool InUse;
+
+			public ZstdCompressionStreamContext()
+			{
+				Zcs = _lib.ZSTD_createCStream();
+
+				InputBuffer = new byte[INPUT_BUFFER_SIZE];
+				InputHandle = GCHandle.Alloc(InputBuffer, GCHandleType.Pinned);
+
+				Input = new()
+				{
+					Ptr = InputHandle.AddrOfPinnedObject(),
+					Size = 0,
+					Pos = 0,
+				};
+
+				OutputBuffer = new byte[OUTPUT_BUFFER_SIZE];
+				OutputHandle = GCHandle.Alloc(OutputBuffer, GCHandleType.Pinned);
+
+				Output = new()
+				{
+					Ptr = OutputHandle.AddrOfPinnedObject(),
+					Size = OUTPUT_BUFFER_SIZE,
+					Pos = 0,
+				};
+
+				InUse = false;
+			}
+
+			~ZstdCompressionStreamContext()
+			{
+				Dispose();
+			}
+
+			private bool _disposed = false;
+
+			public void Dispose()
+			{
+				if (!_disposed)
+				{
+					_lib.ZSTD_freeCStream(Zcs);
+					InputHandle.Free();
+					OutputHandle.Free();
+					GC.SuppressFinalize(this);
+					_disposed = true;
+				}
+			}
+
+			public void InitContext(int compressionLevel)
+			{
+				if (InUse)
+				{
+					throw new InvalidOperationException("Cannot init context still in use!");
+				}
+
+				_lib.ZSTD_initCStream(Zcs, compressionLevel);
+				Input.Size = Input.Pos = Output.Pos = 0;
+				InUse = true;
+			}
+		}
+
+		private sealed class ZstdCompressionStream : Stream
+		{
+			private readonly Stream _baseStream;
+			private readonly ZstdCompressionStreamContext _ctx;
+
+			public ZstdCompressionStream(Stream baseStream, ZstdCompressionStreamContext ctx)
+			{
+				_baseStream = baseStream;
+				_ctx = ctx;
+			}
+
+			~ZstdCompressionStream()
+			{
+				Dispose();
+			}

```

Comment:
Due to the unconditional access of managed fields in the dispose function, this isn't actually safe to do and could crash the entire program.
This should probably either be a `Dispose(false)` or not implement the finalizer at all and hope this always gets disposed correctly.

You should be able to reproduce this failure by slamming
```csharp
for (int i = 0; i < 100; i++)
	new Zstd().CreateZstdCompressionStream(new MemoryStream());
```
somewhere, like in main.